apps/aefate/src/aefa_chain_api.erl

```diff
@@ -208,19 +247,43 @@ contract_fate_code(Pubkey, #state{primop_state = PState} = S) ->
                             {ref, Ref} ->
                                 RefContractPK = aeser_id:specialize(Ref, contract),
                                 {value, RefContract} = aeprimop_state:find_contract_without_store(RefContractPK, PState),
-                                {code, Code} = aect_contracts:code(RefContract),
-                                Code
+                                case aect_contracts:code(RefContract) of
+                                    {ref, RefBad} ->
+                                        error({found_deep_reference, RefBad});
+                                    {code, Code} -> Code
+                                end
                         end,
                     #{ byte_code := ByteCode} = aect_sophia:deserialize(SerCode),
                     try aeb_fate_code:deserialize(ByteCode) of
-                        FateCode -> {ok, {FateCode, VMV}, S#state{primop_state = PState}}
+                        FateCode -> {ok, FateCode, VMV, S0#state{primop_state = PState}}
                     catch _:_ -> error
                     end;
                 _ ->
                     error
             end
     end.
 
+-spec contract_find_final_ref(pubkey(), state()) -> 'error' | {'ok', pubkey(), aect_contracts:vm_version()}.
+contract_find_final_ref(Pubkey, #state{primop_state = PState} = S0) ->
+    case aeprimop_state:find_contract_without_store(Pubkey, PState) of
+        none -> error;
+        {value, Contract} ->
+            case aect_contracts:code(Contract) of
+                {code, _} ->
+                    {ok, Pubkey, aect_contracts:vm_version(Contract)};
+                {ref, Ref} ->
+                    RefContractPK = aeser_id:specialize(Ref, contract),
+                    contract_find_final_ref(RefContractPK, S0)
```

Comment:
Infinite loop in case of cyclic references - Good thing we enforce that we can't create cyclic references :)

Reply:
This code actually prevents cyclic references from happening – unless someone puts in their DB such a one manually (which would be invalid anyways).